query/src/test/java/io/keen/client/java/KeenQueryTest.java

```diff
@@ -542,8 +612,1159 @@ public void testNullResult() throws Exception {
+            Boolean withActors = null;
+            
+            if (stepJson.has(KeenQueryConstants.TIMEFRAME) &&
+                stepJson.has(KeenQueryConstants.TIMEZONE)) {
+                timeframe = new RelativeTimeframe(
+                        stepJson.get(KeenQueryConstants.TIMEFRAME).asText(),
+                        stepJson.get(KeenQueryConstants.TIMEZONE).asText());
+            }
+            else if(stepJson.has(KeenQueryConstants.TIMEFRAME)) {
+                JsonNode timeframeJson = stepJson.get(KeenQueryConstants.TIMEFRAME);
+                if (!timeframeJson.isObject()) {
+                    timeframe = new RelativeTimeframe(timeframeJson.asText());
+                }
+                else {
+                    throw new IllegalStateException(
+                            "Building absolute timeframes isn't supported by this method.");
+                }
+            }
+            
+            if (stepJson.has(KeenQueryConstants.FILTERS)) {
+                JsonNode filterListJson = stepJson.get(KeenQueryConstants.FILTERS);
+                Iterator<JsonNode> filterJsonIterator = filterListJson.iterator();
+                while (filterJsonIterator.hasNext()) {
+                    JsonNode filterJson = filterJsonIterator.next();
+                    if (null == filters) {
+                        filters = new LinkedList<Filter>();
+                    }
+                    filters.add(
+                        new Filter(
+                            filterJson.get(KeenQueryConstants.PROPERTY_NAME).asText(),
+                            stringToFilterOperator(filterJson.get(KeenQueryConstants.OPERATOR).asText()),
+                            filterJson.get(KeenQueryConstants.PROPERTY_VALUE).asText()
+                        )
+                    );
+                }
+            }
+            
+            if (stepJson.has(KeenQueryConstants.INVERTED)) {
+                inverted = stepJson.get(KeenQueryConstants.INVERTED).asBoolean();
+            }
+            
+            if (stepJson.has(KeenQueryConstants.OPTIONAL)) {
+                optional = stepJson.get(KeenQueryConstants.OPTIONAL).asBoolean();
+            }
+            
+            if (stepJson.has(KeenQueryConstants.WITH_ACTORS)) {
+                withActors = stepJson.get(KeenQueryConstants.WITH_ACTORS).asBoolean();
+            }
+            
+            FunnelStep step = new FunnelStep(
+                    stepJson.get(KeenQueryConstants.EVENT_COLLECTION).asText(),
+                    stepJson.get(KeenQueryConstants.ACTOR_PROPERTY).asText(),
+                    timeframe,
+                    filters,
+                    inverted,
+                    optional,
+                    withActors
+            );
+            
+            funnelSteps.add(step);
+        }
+        
+        return funnelSteps;
+    }
+    
+    @Test
+    public void testFunnelWithOnlyRequiredParameters() throws Exception {
```

Comment:
Suggestion: I love to see all this test coverage! But I don't especially like seeing 1000+ line files, even if it's just tests. Might be worth breaking this up into a few separate files, e.g. one for funnel tests, one for multi-analysis tests, one for everything else (as a starting point). Not a ship-blocking requirement but I think it would make navigating through the tests a little easier.